firmware/powerboard2/code/app/powerManagement.c

```diff
@@ -0,0 +1,201 @@
+} powerManagement_state_E;
+
+typedef struct
+{
+    uint32_t currents[POWER_MANAGEMENT_CHANNEL_COUNT];
+    powerManagement_state_E state;
+    uint32_t initTimestamp;
+} powerManagement_data_S;
+
+static powerManagement_data_S powerManagement_data;
+
+void powerManagement_init(void)
+{
+	// GPIO Initialization
+	RCC_AHB1PeriphClockCmd(RCC_AHB1Periph_GPIOC, ENABLE);
+
+	GPIO_InitTypeDef GPIO_InitStruct;
+	GPIO_InitStruct.GPIO_Pin   = _VBATT_ENABLE_GPIO | _5V_ENABLE_GPIO | _12V_ENABLE_GPIO | _16V_ENABLE_GPIO;
+	GPIO_InitStruct.GPIO_Mode  = GPIO_Mode_OUT;
+	GPIO_InitStruct.GPIO_Speed = GPIO_Speed_2MHz;
+	GPIO_InitStruct.GPIO_OType = GPIO_OType_PP;
+	GPIO_InitStruct.GPIO_PuPd  = GPIO_PuPd_NOPULL;
+	GPIO_Init(GPIOC, &GPIO_InitStruct);
+
+	//PD2 is used for giving a pulse to the register
+	GPIO_InitStruct.GPIO_Pin = Pulse_GPIO;
+	GPIO_InitStruct.GPIO_Speed = GPIO_Speed_25MHz;
+	GPIO_Init(GPIOD, &GPIO_InitStruct);
+
+	//Keep PD2 output low after initialization
+	GPIOD->ODR &= ~(Pulse_GPIO);
+
+	powerManagement_data.state = POWER_MANAGEMENT_STATE_INIT;
+	powerManagement_data.initTimestamp = RTOS_getTimeMilliseconds();
+}
+
+void powerManagement_run100ms(void)
+{
+    switch(powerManagement_data.state)
+    {
+        case POWER_MANAGEMENT_STATE_INIT:
+        {
+            // Waiting for 1s to pass
+        	if(RTOS_getTimeElapsedMilliseconds(powerManagement_data.initTimestamp) > DELAY_BEFORE_POWER_ON_MS)
+            {
+                // Turn on all the power
+                powerManagement_setState(POWER_MANAGEMENT_CHANNEL_VBATT, true);
+                powerManagement_setState(POWER_MANAGEMENT_CHANNEL_5V, true);
+                powerManagement_setState(POWER_MANAGEMENT_CHANNEL_12V, true);
+                powerManagement_setState(POWER_MANAGEMENT_CHANNEL_16V, true);
+
+                // PD2 pulse: transition from low to high to write to the register
+                GPIOD->ODR |= Pulse_GPIO;
+
+                powerManagement_data.state = POWER_MANAGEMENT_STATE_RUN;
+            }
+            break;
+        }
+        case POWER_MANAGEMENT_STATE_RUN:
+        {
+        	//PD2 pulse: set PD2 to low at the beginning of run state
+        	GPIOD->ODR &= ~(Pulse_GPIO);
+
+        	protocol_allMessages_U message;
+            // Obtain Trident's instruction about power
+            messageHandler_getMessage(MESSAGE_HANDLER_RX_CHANNEL_POWER_ENABLE, &message, NULL);
+
+            // Control the power according to Trident's orders
+            powerManagement_setState(POWER_MANAGEMENT_CHANNEL_VBATT, message.TRIDENT_powerEnable.VBattPowerEnable);
+            powerManagement_setState(POWER_MANAGEMENT_CHANNEL_5V, message.TRIDENT_powerEnable._5VPowerEnable);
+            powerManagement_setState(POWER_MANAGEMENT_CHANNEL_12V, message.TRIDENT_powerEnable._12VPowerEnable);
+            powerManagement_setState(POWER_MANAGEMENT_CHANNEL_16V, message.TRIDENT_powerEnable._16VPowerEnable);
+
+
+			
+			/*Implement this after housingMonitor code is done
```

Comment:
Can you implement this now. It looks like your housing monitor code is complete.